mm2src/coins/utxo/rpc_clients/electrum_rpc/connection.rs

```diff
@@ -0,0 +1,872 @@
+
+        let response = match response {
+            ElectrumRpcResponseEnum::SingleResponse(single) => JsonRpcResponseEnum::Single(single),
+            ElectrumRpcResponseEnum::BatchResponses(batch) => JsonRpcResponseEnum::Batch(batch),
+            ElectrumRpcResponseEnum::SubscriptionNotification(req) => {
+                // NOTE: Sending a script hash notification is handled in it's own event handler.
+
+                // FIXME: What is this used for? Note that the id is the method name in this case (two similar id will collide),
+                // but this isn't a response to any request anyways, this is a notification, and we forwarded using the
+                // scripthash_notification_sender above already.
+                JsonRpcResponseEnum::Single(JsonRpcResponse {
+                    id: req.method.clone(),
+                    jsonrpc: "2.0".into(),
+                    result: req.params[0].clone(),
+                    error: Json::Null,
+                })
+                //return; // BLOCKCHAIN_HEADERS_SUB_ID wasn't handled, just returned.
+                //also, you might want to check the notification type to print an error message if it's not expected
+            },
+        };
+
+        // the corresponding sender may not exist, receiver may be dropped
+        // these situations are not considered as errors so we just silently skip them
+        let pending = self.responses.lock().await.remove(&response.rpc_id());
+        if let Some(tx) = pending {
+            tx.send(response).ok();
+        }
+    }
+
+    /// Process a bulk response from the electrum server.
+    ///
+    /// A bulk response is a response that contains multiple JSONRPC responses.
+    async fn process_electrum_bulk_response(
+        &self,
+        bulk_response: &[u8],
+        event_handlers: &Vec<Box<SharableRpcTransportEventHandler>>,
+    ) {
+        // We should split the received response because we can get several responses in bulk.
+        let responses = bulk_response.split(|item| *item == b'\n');
+
+        for response in responses {
+            // `split` returns empty slice if it ends with separator which is our case.
+            if !response.is_empty() {
+                let raw_json: Json = match json::from_slice(response) {
+                    Ok(json) => json,
+                    Err(e) => {
+                        error!("{}", e);
+                        return;
+                    },
+                };
+                self.process_electrum_response(raw_json, event_handlers).await
+            }
+        }
+    }
+
+    /// Starts the connection loop that keeps an active connection to the electrum server.
+    ///
+    /// This will first try to connect to the server and use that connection to query its version.
+    /// If version checks succeed, the connection will be kept alive, otherwise, it will be dropped.
+    #[cfg(not(target_arch = "wasm32"))]
+    pub async fn establish_connection_loop(
```

Comment:
the function is quite long. can it be broken down into smaller, more manageable pieces and incorporate with the `wasm32` version

Reply:
Agreeeed!
I wanted to merge `wasm` within but found it it would be way way longer.
Will try my luck splitting it up.

Reply:
Done here [3fb958b](https://github.com/KomodoPlatform/komodo-defi-framework/pull/1966/commits/3fb958b093cdb4b57b33e876a475ab38c02634a0).

Still big though. But i find it logically easy to traverse (or maybe cuz i spent nights with it already xD)